### PR TITLE
Implement DAP_SWJ_CLOCK command

### LINKED ISSUE #4: [CMSIS-DAP] Implement DAP_SWJ_CLOCK command
Create an implementation for CMSIS-DAP DAP_SWJ_CLOCK command.

### PR DESCRIPTION
**Pull request by efk-certi**
close #4

### CODE REVIEW

File: jocd/src/br/org/certi/jocd/dapaccess/CmsisDapProtocol.java [hunk 1]
```diff
@@ -194,8 +199,29 @@ public byte transferConfigure(byte idleCycles, int waitRetry, int matchRetry)
...
 
-  public byte setSWJClock(int freq) {
-    // TODO
-    return 0x00;
+  public int setSWJClock() throws DeviceError, CommandError {
+    return setSWJClock(1000000);
+  }
+
+  public byte setSWJClock(int clock) throws DeviceError, CommandError {
+    byte[] cmd = new byte[2];
```
Comment (inm-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/CmsisDapProtocol.java]: This byte array has 5 elements, so... new byte[5]

File: jocd/src/br/org/certi/jocd/dapaccess/Command.java [hunk 2]
```diff
@@ -49,4 +54,198 @@ public Command(int size) {
...
+    }
+
+    // Check for count mismatch after checking for DAP_TRANSFER_FAULT
+    // This allows TransferFaultError or TransferTimeoutError to get
+    // thrown instead of TransferFaultError
+    if (data[1] != this.readCount + this.writeCount) {
+      throw new TransferError();
+    }
+
+    int arraySize = 3 + 4 * this.readCount;
```
Comment (inm-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/Command.java]: Ins't this supposed to be 4 * this.readCount - 3?
I think you are taking out the first 3 elements, right?

File: jocd/src/br/org/certi/jocd/dapaccess/Command.java [hunk 3]
```diff
@@ -49,4 +54,198 @@ public Command(int size) {
...
+
+    // Check for count mismatch after checking for DAP_TRANSFER_FAULT
+    // This allows TransferFaultError or TransferTimeoutError to get
+    // thrown instead of TransferFaultError
+    int transferCount = data[1] | (data[2] << 8);
+    if (transferCount != this.readCount + this.writeCount) {
+      throw new TransferError();
+    }
+
+    int arraySize = 4 + 4 * this.readCount;
```
Comment (inm-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/Command.java]: Same here...

File: jocd/src/br/org/certi/jocd/dapaccess/Command.java [hunk 4]
```diff
@@ -49,4 +54,198 @@ public Command(int size) {
...
+    assert this.dataEncoded == true;
+    if (this.blockAllowed) {
+      data = this.decodeTransferBlockData(data);
+    } else {
+      data = this.decodeTransferData(data);
+    }
+    return data;
+  }
+
+
```
Comment (inm-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/Command.java]: Extra line

File: jocd/src/br/org/certi/jocd/dapaccess/CmsisDapProtocol.java (hunk 1 above)
Comment (efk-certi) [reply jocd/src/br/org/certi/jocd/dapaccess/CmsisDapProtocol.java]: Fixed

File: jocd/src/br/org/certi/jocd/dapaccess/Command.java [hunk 5]
```diff
@@ -16,6 +16,11 @@
 package br.org.certi.jocd.dapaccess;
 
 import android.util.Log;
+import br.org.certi.jocd.dapaccess.CmsisDapCore.CommandId;
+import br.org.certi.jocd.dapaccess.dapexceptions.TransferError;
+import br.org.certi.jocd.dapaccess.dapexceptions.TransferFaultError;
+import br.org.certi.jocd.dapaccess.dapexceptions.TransferTimeoutError;
+import java.util.List;
```
Comment (ghn-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/Command.java]: The import of project files and system files should be in a different blocks.

File: jocd/src/br/org/certi/jocd/dapaccess/DapAccessCmsisDap.java [hunk 6]
```diff
@@ -24,31 +24,33 @@
...
+import java.util.LinkedList;
 import java.util.List;
 
 public class DapAccessCmsisDap {
 
   // Logging
   private static final String TAG = "DapAccessCmsisDap";
 
+  //CMSIS-DAP values
+  public static byte READ = 1 << 1;
```
Comment (ghn-certi) [inline jocd/src/br/org/certi/jocd/dapaccess/DapAccessCmsisDap.java]: This value could be declared as final.

**Pull request merged**